src/storage-controller/src/lib.rs

```diff
@@ -1705,76 +1709,130 @@ where
 
     async fn record_frontiers(
         &mut self,
-        external_frontiers: BTreeMap<(GlobalId, ReplicaId), Antichain<Self::Timestamp>>,
+        external_frontiers: BTreeMap<
+            GlobalId,
+            (Antichain<Self::Timestamp>, Antichain<Self::Timestamp>),
+        >,
     ) {
-        // Make `replica_id` optional, to account for storage objects that are not installed on
-        // replicas.
-        let mut frontiers = BTreeMap::new();
-        let mut external_ids = HashSet::with_capacity(frontiers.len());
-        for ((object_id, replica_id), frontier) in external_frontiers {
-            frontiers.insert((object_id, Some(replica_id)), frontier);
-            external_ids.insert(object_id);
+        let mut frontiers = external_frontiers;
+
+        // Enrich `frontiers` with storage frontiers.
+        for (object_id, collection) in self.active_collections() {
+            let since = collection.read_capabilities.frontier().to_owned();
+            let upper = collection.write_frontier.clone();
+            frontiers.insert(object_id, (since, upper));
+        }
+        for (object_id, export) in self.active_exports() {
+            // Exports cannot be read from, so their `since` is always the empty frontier.
+            let since = Antichain::new();
+            let upper = export.write_frontier.clone();
+            frontiers.insert(object_id, (since, upper));
         }
 
+        let mut updates = Vec::new();
+        let mut push_update =
+            |object_id: GlobalId,
+             (since, upper): (Antichain<Self::Timestamp>, Antichain<Self::Timestamp>),
+             diff: Diff| {
+                let read_frontier = since
+                    .into_option()
+                    .map_or(Datum::Null, |ts| Datum::MzTimestamp(ts.into()));
+                let write_frontier = upper
+                    .into_option()
+                    .map_or(Datum::Null, |ts| Datum::MzTimestamp(ts.into()));
+                let row = Row::pack_slice(&[
+                    Datum::String(&object_id.to_string()),
+                    read_frontier,
+                    write_frontier,
+                ]);
+                updates.push((row, diff));
```

Comment:
If we kept read and write frontier separate, we might be able to push the timestamp into the diff field. Just mentioning this so we don't forget that's also an option, but out of scope for this PR.